File: android/src/main/java/com/margelo/nitro/nitrospeech/recognizer/logic/RepeatingFilter.kt
package com.margelo.nitro.nitrospeech.recognizer.logic

object RepeatingFilter {
  /**
   * Drops consecutive duplicate words while always keeping number-containing tokens.
   * A stable prefix of 10- tokens is preserved; only the last 10 tokens are filtered.
   */
  fun apply(text: String): String {
    var words = text.split(Regex("\\s+")).filter { it.isNotBlank() }
    if (words.isEmpty()) {
      return ""
    }

    val joiner = StringBuilder()

    if (words.size >= 10) {
      joiner.append(words.take(words.size - 9).joinToString(" "))
      words = words.takeLast(10)
    } else {
      joiner.append(words.first())
    }

    for (i in words.indices) {
      if (i == 0) continue
      if (Regex("\\d+").containsMatchIn(words[i])) {
        joiner.append(" ").append(words[i])
        continue
      }
      if (words[i] == words[i - 1]) continue
      joiner.append(" ").append(words[i])
    }
    return joiner.toString()
  }
}

File: ios/Logic/RepeatingFilter.swift
import Foundation

enum RepeatingFilter {
    /// Drops consecutive duplicate tokens while keeping numbers and a stable prefix.
    static func apply(_ text: String) -> String {
        var subStrings = text.split { $0.isWhitespace }.map { String($0) }
        if #available(iOS 16.0, macOS 13.0, *) {
            var shift = 0
            while shift < subStrings.count, !subStrings[shift].contains(/\w+/) {
                shift += 1
            }
            if shift > 0, shift < subStrings.count {
                subStrings = Array(subStrings.suffix(subStrings.count - shift))
            }
        }
        guard !subStrings.isEmpty else { return text }

        var joiner = ""
        if subStrings.count >= 10 {
            joiner = subStrings.prefix(subStrings.count - 9).joined(separator: " ")
            subStrings = Array(subStrings.suffix(10))
        } else {
            joiner = subStrings.first ?? ""
        }
        for i in subStrings.indices {
            if i == 0 { continue }
            if #available(iOS 16.0, macOS 13.0, *), subStrings[i].contains(/\d+/) {
                joiner += " \(subStrings[i])"
                continue
            }
            if subStrings[i] == subStrings[i - 1] { continue }
            joiner += " \(subStrings[i])"
        }
        return joiner
    }
}
